// cypress-json-html-reporter/dist/esm/lib/CypressJSONReporter.js
import fs from 'fs';;
import path from 'path';;
import Mocha from 'mocha';;
const {
  EVENT_RUN_END,
  EVENT_TEST_PASS,
  EVENT_TEST_FAIL,
  EVENT_TEST_PENDING
} = Mocha.Runner.constants;

/**
 * Custom Cypress JSON Reporter that generates a report compatible
 * with wdio-json-html-reporter generators.
 *
 * Options:
 * - outputFile: the path where the JSON file is to be written.
 * - screenshotOption: 'OnFailure' to embed screenshots, 'No' to skip.
 */
class CypressJSONReporter extends Mocha.reporters.Base {
  constructor(runner, options) {
    super(runner, options);

    // Parse reporter configurations passed from Cypress
    const reporterOptions = options.reporterOptions || {};
    this.options = Object.assign({
      stdout: true,
      screenshotOption: 'No',
      history: 5,
      historyPath: null,
      outputFile: 'test-report.json'
    }, reporterOptions);
    this.testResults = [];
    this.testResultUids = new Set();

    // Shared execution state
    this.executionId = Date.now().toString();
    this.executionStartTime = new Date();

    // Path to the shared screenshot map written by the Cypress plugin event
    const outputDir = path.dirname(this.options.outputFile);
    this.screenshotMapFile = path.join(outputDir, '.screenshot-map.json');

    // Ensure output directory exists
    if (!fs.existsSync(outputDir)) {
      fs.mkdirSync(outputDir, {
        recursive: true
      });
    }
    runner.on(EVENT_TEST_PASS, test => {
      this.addTestResult(test, 'PASSED');
    });
    runner.on(EVENT_TEST_FAIL, (test, err) => {
      this.addTestResult(test, 'FAILED', err);
    });
    runner.on(EVENT_TEST_PENDING, test => {
      this.addTestResult(test, 'SKIPPED');
    });
    runner.on(EVENT_RUN_END, () => {
      this.writeJSONReport();
    });
  }

  /**
   * Reads the browser info written by the cypress.config.js before:run event.
   */
  readBrowserInfo(outputDir) {
    const pathsToTry = [path.join(outputDir, '.browser-info.json'), path.join(process.cwd(), 'reports', '.browser-info.json')];
    for (const p of pathsToTry) {
      if (fs.existsSync(p)) {
        try {
          const info = JSON.parse(fs.readFileSync(p, 'utf8'));
          if (info && info.browserName) {
            return info.browserName.charAt(0).toUpperCase() + info.browserName.slice(1);
          }
        } catch (e) {
          // Ignore
        }
      }
    }
    return process.env.CYPRESS_BROWSER || 'Desktop Browser';
  }

  /**
   * Reads the screenshot map written by the cypress.config.js after:screenshot event.
   * Returns a map of { key => absolutePath }.
   */
  readScreenshotMap() {
    if (fs.existsSync(this.screenshotMapFile)) {
      try {
        return JSON.parse(fs.readFileSync(this.screenshotMapFile, 'utf8'));
      } catch (e) {
        return {};
      }
    }
    return {};
  }

  /**
   * Converts an image file to a base64 data URL.
   * Returns empty string if the file cannot be read.
   */
  imageToDataUrl(filePath) {
    try {
      const ext = path.extname(filePath).toLowerCase().replace('.', '');
      const mimeType = ext === 'jpg' || ext === 'jpeg' ? 'image/jpeg' : 'image/png';
      const data = fs.readFileSync(filePath);
      return `data:${mimeType};base64,${data.toString('base64')}`;
    } catch (e) {
      console.warn(`[CypressJSONReporter] Could not read screenshot: ${filePath}`);
      return '';
    }
  }
  addTestResult(test, status, err = null) {
    const date = new Date();
    // Time portion in HH:MM:SS format
    const timestamp = date.toISOString().split('T')[1].split('.')[0];

    // Create unique ID based on test properties
    const testTitle = test.title || 'Unknown Test';
    const suiteName = test.parent ? test.parent.title : 'Default Suite';
    const uid = `${testTitle}-${timestamp}-${Date.now()}`;

    // Handle errors (Cypress/Mocha provides an Error object err)
    let errors = [];
    if (err) {
      errors.push({
        message: this.sanitizeErrorMessage(err.message, true),
        stack: this.sanitizeErrorMessage(err.stack, true)
      });
    } else if (test.err) {
      errors.push({
        message: this.sanitizeErrorMessage(test.err.message, true),
        stack: this.sanitizeErrorMessage(test.err.stack, true)
      });
    }

    // Look up screenshot from the shared map written by the cypress.config.js plugin event.
    // Rules: PASSED → screenshot, FAILED → screenshot, SKIPPED → none.
    // Key format:
    //   passed  → "SuiteName -- TestName (passed)"  (from cy.screenshot in support/e2e.js)
    //   failed  → "SuiteName -- TestName (failed)"  (Cypress auto-screenshot on failure)
    let screenshotPath = '';
    if (status !== 'SKIPPED' && this.options.screenshotOption !== 'No') {
      const screenshotMap = this.readScreenshotMap();
      const suffix = status === 'PASSED' ? '(passed)' : '(failed)';
      const expectedKey = `${suiteName} -- ${testTitle} ${suffix}`;
      if (screenshotMap[expectedKey]) {
        screenshotPath = this.imageToDataUrl(screenshotMap[expectedKey]);
      } else {
        // Fallback: find any key that includes the test title and the right status suffix
        const matchedKey = Object.keys(screenshotMap).find(k => k.includes(testTitle) && k.includes(suffix));
        if (matchedKey) {
          screenshotPath = this.imageToDataUrl(screenshotMap[matchedKey]);
        }
      }
    }

    // Cypress does not expose a straight forward `browser.getLogs` stream to the reporter.
    let browserConsoleLogs = [];
    let specConsoleLogs = []; // Handled separately or intercepted if users apply cy.on plugins

    if (!this.testResultUids.has(uid)) {
      this.testResults.push({
        uid,
        timestamp,
        suiteName,
        testName: testTitle,
        status,
        errors,
        screenshot: screenshotPath,
        browserConsoleLogs,
        specConsoleLogs
      });
      this.testResultUids.add(uid);
    }
  }
  writeJSONReport() {
    const executionEnd = new Date();
    const totalTimeMinutes = ((executionEnd - this.executionStartTime) / 60000).toFixed(2);
    const outputDir = path.dirname(this.options.outputFile);
    const browserName = this.readBrowserInfo(outputDir);
    const metadata = {
      executionId: this.executionId,
      browserName,
      executionStartTime: this.executionStartTime.toUTCString(),
      executionEndTime: executionEnd.toUTCString(),
      totalTimeInMinutes: totalTimeMinutes
    };
    const report = {
      metadata,
      testResults: this.testResults
    };
    const fileTimestamp = this.formatDateForFilename(new Date());
    const fileName = `test-report-${fileTimestamp}.json`;
    const outputFile = path.join(outputDir, fileName);
    try {
      fs.writeFileSync(outputFile, JSON.stringify(report, null, 2));
      console.log(`JSON report successfully written to ${outputFile}`);
    } catch (writeErr) {
      console.error(`Failed to write JSON report to ${outputFile}:`, writeErr);
    }
  }
  formatDateForFilename(date) {
    return date.toUTCString().replace(/,/g, '').replace(/:/g, '-').replace(/\s+/g, '_');
  }
  sanitizeErrorMessage(errorMessage, full = false) {
    if (!errorMessage) return '';
    const sanitized = errorMessage.replace(/[\u001b\u009b]\[\d{1,2}(;\d{1,2})?(m|K)/g, '').trim();
    return full ? sanitized : sanitized.split('\n')[0];
  }
}
export default CypressJSONReporter;
